String mapi ~2.5x perf improvement (v2)

abelbraaksma opened the pull request:
See for details of timings that motivated this improvement: https://github.com/dotnet/fsharp/issues/9390#issuecomment-642985879. 

TLDR: using a `char []` improves ~2.1x, ~then 2x unrolling brings this to \~2.5x~ (edit: no unrolling will be done, the JIT should do that). In other words, if it took 1 second before, it now takes 0.4 seconds.

Dependent on: https://github.com/dotnet/fsharp/pull/9481.

cartermp commented on src/fsharp/FSharp.Core/string.fs:
```diff
@@ -43,13 +43,24 @@ namespace Microsoft.FSharp.Core
-                str |> iteri (fun i c -> res.Append(f.Invoke(i, c)) |> ignore)
-                res.ToString()
+                let result = str.ToCharArray()
+                let f = OptimizedClosures.FSharpFunc<_,_,_>.Adapt mapping
```
nit: the change to remove the parentheses isn't necessary here

abelbraaksma replied:
@cartermp Thanks, fixed here: https://github.com/dotnet/fsharp/pull/9482/commits/3744f5301c08149f3e1ed5fc8bc159c124a72898 (which also fixes the CI errors).

I'm no fan of parens, hence I removed it ;)

abelbraaksma commented:
Hmm, the CI errors are odd:

>    at NUnit.Framework.Assert.AreEqual(Object expected, Object actual, String message) in /home/vsts/work/1/s/tests/FSharp.Core.UnitTests/NUnitFrameworkShims.fs:line 93
>   at FSharp.Core.UnitTests.Collections.StringModule.MapI() in /home/vsts/work/1/s/tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/StringModule.fs:line 85

That's this test:

```f#
[<Test>]
member this.MapI() =
    let e1 = String.mapi (fun i c -> char(int c + i)) "foo"
    Assert.AreEqual("fpq", e1)

    let e2 = String.mapi (fun i c -> c) null 
    Assert.AreEqual("", e2)
```

And yes, these were wrong, incorrectly copied & changed the code from the BDN project. I've added more tests since we're now unrolling by `% 2`, which warranted adding the corners for `0`, `1`, `2` and I've added a few general ones to ensure the mapping function isn't called when it shouldn't: https://github.com/dotnet/fsharp/pull/9482/commits/4d15697056823658f3683302a885a12b3dba6107

abelbraaksma commented:
Another sourcelink fail? 

> C:\Users\VssAdministrator\.nuget\packages\microsoft.sourcelink.common\1.1.0-beta-20206-02\build\InitializeSourceControlInformation.targets(3,81): error MSB4022: The result "" of evaluating the value "$(_MicrosoftSourceLinkCommonAssemblyFile)" of the "AssemblyFile" attribute in element <UsingTask> is not valid.

Not sure what causes this, doesn't seem related.

abelbraaksma commented:
All's green now, newly added tests passed.

KevinRansom commented on src/fsharp/FSharp.Core/string.fs:
```diff
@@ -40,13 +43,25 @@ namespace Microsoft.FSharp.Core
-                res.ToString()
+
+                // x2 unrolled loop gives 10-20% boost, overall 2.5x SB perf
+                let mutable i = 0
```
using ToCharArray() is good, however, I would rather the Jit team fix their code.  Also, on the coreclr and desktop FSharp.Core is often used when ngened/crossgened

abelbraaksma replied:
> using ToCharArray() is good, however, I would rather the Jit team fix their code

@KevinRansom I think the 2nd half of that sentence is about the unrolling, right? I mean, you're not opposing this change to use `ToCharArray()` in principal (it has no bearing with how JIT does things)? 

After our discussion yesterday, I'm now convinced that unrolling should only be done if we can prove it's indeed better (I believe this to be true for the `String.iter` case, the benefit is also more dramatic there). It looks like more modern CPU's than mine have an advantage through:

* This [insanely detailed SO answer](https://stackoverflow.com/a/39940932/111575) on loop optimizations per architecture shows that efficient use of the LSD (Loop Stream Detector) is highly dependent on the size of the loop.
* Server architectures often have a larger near-cache (L0/L1) and multi phys procs use the QPI bus, which all leads to different behaviors (I suspect the latter to be more susceptible to this optimization, as experiments have shown)
* Larger inputs have bigger advantages. Your test uses 22 chars. However, even larger can show detrimental effects as a result of not fitting into the CPU cache:
![image](https://user-images.githubusercontent.com/16015770/85153844-fb397600-b256-11ea-971a-406130008b8a.png)
* The advantage of unrolling is ~10-20%, I agree this is not worth it in this case. For the same as above, this is the non-unrolled version of this proposal, i.e., `fasterMap`:
![image](https://user-images.githubusercontent.com/16015770/85154002-2b811480-b257-11ea-90fa-d40f151eb67a.png)

About the JIT fixing unrolling, it is unlikely to help us here:
* JIT needs to know statically what the number of loops is (n/a here)
* RyuJIT currently doesn't do this, and efforts since 2015/2017 have proven very hard: https://github.com/dotnet/runtime/issues/8107
* Legacy JIT, surprisingly, does do this https://aakinshin.net/posts/unrolling-of-small-loops-in-different-jit-versions/ (same article shows what (not) happens for other JITs)
* Even if statically known, the number of loops need to be a multiple of N, where N is the number of unrolls. There's effort underway to lift this requirement, but it may be many years, if ever.

Actual unrolling at IL level could make _some_ sense, but the code created when doing it at F# language level is far from optimal. C# can do this better, but for a good use case for all this work, we need good proof of the generated x64 assembly.

All in all, I agree to update this PR to remove the unrolling in this case, as we've seen it only gets better on _some_ processors.

PS: I've investigate the disassembly you sent me, I'll get back to you on that. It's certainly interesting.

abelbraaksma replied:
@KevinRansom, this has been resolved: https://github.com/dotnet/fsharp/pull/9482/commits/3936ad035138c2a90a27061b415dacc1d6dd95f1

I also removed "foo", which you requested elsewhere, I believe ;).

abelbraaksma commented:
@KevinRansom, the unrolling was removed, as discussed, all requested changes implemented and green. Good to go?

abelbraaksma commented:
@cartermp, you merged master into this, I think this is now ready?

cartermp commented:
This will require a re-review by @KevinRansom to merge

abelbraaksma commented:
/cc @kevinransom

abelbraaksma commented:
Thanks @kevinransom!

KevinRansom commented:
@abelbraaksma , happy to be of service.
